### ISSUE TITLE
BLT Project Moving Off of Features to Config-Split

### ISSUE THREAD

**Issue by mgamshad** [acquia/blt#3754]
**I want to...**
This is also more of a general BLT question related to CMI:
I want to move our Acquia ACSF multisite BLT 9.2.x application from Features to use core CMI and config-split.

**It's not working because...**
Here is the sequence that I thought may work:
- start with a local build of a reference site pulled from ACSF (includes blt setup and blt sync:refresh)
- enable config-split and config-filter modules
- export the configuration
- modify blt.yml to use config-split for CM
- Remove any *.features.yml file in any module that used Features (not necessarily all the config files)
- Validate sandbox can be rebuilt

Question on validation: what should be the `blt` commands to rerun to make sure the sandbox can be rebuilt ok, but given that `blt setup` starts from scratch I would assume that wouldn't be part of the steps!!?  Or does this move off of Features to config-split requires modifications done to the upstream source reference site (i.e. the Prod/live site) first and then work backwards?

A second question is about the UUIDs in the exported files.  If I strip the UUIDs and the _core.default_config_hash attributes, and try to run `blt drupal:config:import` (as a validation of the use of core CMI and config-split, this fails complaining about UUID mismatches.

If I don't touch the UUID, of if I just set the UUID to the same value in the exported `system.site.yml`, then the import via `blt drupal:config:import` is successful.

However, given that this is an application code for a multisite setup in ACSF (which is slightly different than "traditional" D8 multisite), what should be the fate of the UUID issue above?  I can't imagine that they would need to be preserved!!!

Any insight on the above questions, or any input on the sequence of steps to achieve this goal of moving a BLT 9.x ACSF project off Features to core CMI and config-split would be appreciated.

**System information**
* Operating system type: macOS
* Operating system version: 10.14.5
* BLT version: 9.2.8

Thanks,
mohsen

**Comment by danepowell** [comment]
I don't think you're going to have an easy time converting to config split by reusing the features-generated config files in place. Features modifies those files by doing things like removing UUIDs, and additionally BLT 10 doesn't natively handle turning splits on/off the way you might with features. BLT 10 only supports site and environment splits, which are architecturally pretty different from Features.

This is documented here, which might be instructive for you: https://docs.acquia.com/blt/developer/config-split/

Additionally, if you want to use config splits on ACSF I would strongly recommend upgrading to BLT 10 first. Support for both ACSF and config management are much more mature in BLT 10.

On the issue of UUIDs, you definitely want to preserve UUIDs in your config files. Otherwise when you import config, Drupal will generate a random UUID which will generally not match the UUID in active storage, causing a failure. This is documented in part here: https://www.drupal.org/project/lightning/issues/2870864

Finally, note that the _site_ UUID and _individual config file_ UUIDs are totally separate beasts, used for different purposes. BLT automatically rewrites the site UUID to match what's in config so that config imports don't fail, but managing UUIDs for individual field/entity definitions is up to you.

**Comment by mgamshad** [comment]
The configuration we manage in Features are all primarily site and environment ones, i.e. we are not managing any esoteric configuration via Features.  However, given that we are on ACSF, i.e. being multi-site, we need a notion of "platform" level configuration, which logically would be in the default sync directory.  Any site-specific configuration can then be managed out of "../config/sites/..." folders.  

So, the question remains as to how can we create "platform"-level configuration in the default sync-directory if those exported configs contain site UUIDs?  The Acquia/BLT config-split document you shared does not discuss ACSF, so not quite sure how to proceed there?  

(Point taken on managing individual field/entity UUID management).

Finally, the move to BLT 10.x seems to be required, and so will add that as a dependency.

Thanks,
Mohsen

**Comment by danepowell** [comment]
That you are referring to "site UUIDs" (as if there are multiple) makes me think that there's something we aren't on the same page about here. There should only be a single site UUID in a given configuration set, that's the one stored in `system.site.yml`. When you run `blt setup` or `blt drupal:config:import`, BLT helps by automatically setting your site UUID to match what's in config: https://github.com/acquia/blt/blob/53649678d4c58dc182bb93f15f096a2108a73a66/src/Robo/Commands/Setup/ConfigCommand.php#L83

Thus as long as you are using BLT commands, it shouldn't matter whether you are in a single-site or multi-site environment.

Any other UUIDs in config yml files are object-level UUIDs and need to be handled independently, but do generally need to be present to avoid UUID mismatch issues.

**Comment by mgamshad** [comment]
Check on the understanding the site UUID vs other UUID instances, so we're on the same page there.  What's not clear right now is how we would migrate off of Features, and use core CMI and config-split, to manage configuration for **all** of the sites on an ACSF stack?  I am assuming that we don't have to create site-specific config-splits for all the sites, and not have a way to identify platform-wide configuration?

**Comment by mgamshad** [comment]
Dane, just saw this: [Acquia Connector Issue](https://www.drupal.org/project/acquia_connector/issues/2968641#comment-12657050), and essentially my config-split questions here about multi-site and UUIDs, speaks to your comment in the issue above where you state: 

> 2) Using Config Split, create empty configuration splits for each of your remote environments if you haven't already.
> 3) Export acquia_connector.settings.yml from each environment and place into the corresponding split. Ensure the keys spi.site_uuid and spi.site_machine_namedo not exist in acquia_connector.settings.yml. This is especially critical on multisite / ACSF installs, because the site UUID and machine name are different per site.

In my case I was asking whether the removal of the site UUID should be done to all of the config files that are created when we do an export to the default config folder to create a "platform" set of configs that are common to all of the sites?  

Thanks

**Comment by mgamshad** [comment]
One more note:  I just was re-testing the setup and did notice that in the default exported configurations, the UUID field there is not the same, which make me now wonder whether there may be a confusion regarding the UUIDs.  The UUID I was referring to in this thread is the very first entry in each of the config files.  Given that they are all different then maybe that's not the site UUID?  

When using Features to manage a configuration, we would use the Export command from "Administration/Configuration  Development/Synchronize" and would strip out the UUID on the first line before committing the YML file to code.  And that was the basis for my initial question, since I was seeing the regular CMI also emit that UUID field into each config file. 

I guess this adds more to the workflow question for an ACSF multisite configuration management.  Anyways, hope this helps a bit.

**Comment by mgamshad** [comment]
Yup, you were correct, I was incorrectly thinking that the UUIDs in the various config yml files were the site's UUID, but in fact as you stated the site UUID is in the `system.site.yml` file.  This was made clear by noting that the UUIDs in the various config yml files were different from each other.

Two questions then: 

1) For our multi-site ACSF platform, given that you mention how BLT 

> helps by automatically setting your site UUID to match what's in config...Thus as long as you are using BLT commands, it shouldn't matter whether you are in a single-site or multi-site environment.

An existing BLT recipe provides splits for the various environments and BLT documentation also mentions support for site splits.  However, it's not clear what the BTL workflow/process should be to manage a multisite setup on ACSF.  For a multisite platform on ACSF, there usually a whole set of configurations that are platform-wide, some configs that are environment specific, and then site-specific configs.  Does BLT have an approach in handling this platform-wide configuration using core CMI and splits, or do we need to create a specific split that is exclusively used to manage the common configuration?  

2) Given the following comment:

> Any other UUIDs in config yml files are object-level UUIDs and need to be handled independently, but do generally need to be present to avoid UUID mismatch issues.

Can you expand on what you mean by handling these object-level UUIDs "independently"?  Again, given the above ACSF platform setup, what's the impact on any handling of these UUIDs on BLT commands and procedures?

Thanks.

**Comment by mgamshad** [comment]
Looking here: [Example Site Split Scenario](https://support.acquia.com/hc/en-us/articles/360024009393) it states that you can set up the splits for a multi-site setup such that the:

> default configuration (lives at config/default) to be shared between all sites

But still not clear how we would deal with the object level UUIDs and how to configure the site split configs incorporated into a BLT workflow?

**Comment by mikemadison13** [comment]
I've never had to strip UUIDs out of objects in this way before in multisite. When you are doing syncs from config, remember that you are using one codebase. So the UUIDs from the "codebase" should be consistent across your platform. I don't think you'll have to do any stripping.

**Comment by mgamshad** [comment]
Thanks.  We stripped them when managing single exports into our Features configs, but I guess Features was managing put that back in!  So, for an ACSF multi-site, would the following workflow be correct (questionable steps noted as such):

- Start by building a local sandbox from a "reference" (any would be correct I am assuming) ACSF site.  We typically use `blt setup` and `blt sync:refresh` to complete this task.
- Set up environment splits by executing:  `blt recipes:config:init:splits `
- Then create the environment splits in the UI
- Export out the entire configuration using `drush cex`
- **Question**: do I need to run the config-split exports to get the splits out as well, i.e. `drush csex` or will the initial export also take care of the config splits?
- **Question**: do I now create config/sites/[site-name] folder for each of the sites we have in ACSF?  I recall that the name must match the name in ACSF, where is the source for that?
- I am assuming that at this time if I have a site-specific config that I can note that in that site's split.
- Set the blt.cm-strategy to config-split.

At this point I am not sure what `blt` commands would be needed to run to make sure that the configuration management, as setup per above, now works and can be validated.  Any input here would be appreciated.

**Comment by mgamshad** [comment]
Not stripping the UUIDs does not seem to be working.  Ignoring any split issues, I am having issues moving from the project BLT using config-split instead of Features.  This is what I have:

- built a local sandbox based on a remote ACSF site
- exported the configuration (`drush cex`)
- changed `blt cm.stragety` from `features` to `config-split`
- and now to test a new site install with the above changes, I ran `blt setup` 

And the result was that I got an error regarding existing entities.  Removing the object level UUIDs did clear that error.  But I had to re-instate the `system.site.yml` UUID, and have it match the upstream remote ACSF site, as I was getting errors related to mismatched site UUIDs.  

But now I get the following error : 

> The import failed due to the following reasons:                              
Cannot change the install profile from [name]  to lightning once Drupal is installed.  Unable to uninstall the [name] profile since it is the main install profile.  

I am not sure if something is over writing the profile name during the sandbox site install, but if I change `core.extension.yml` to have the profile name from the error message then the error is cleared.

Comments?  Thanks.

**Comment by mgamshad** [comment]
On thing I noticed is that when BLT's `cm.strategy = config-split`, building the site using `blt setup` there are some configuration differences due to the fact that BLT is building for a `dev` environment. So I may try to load the actual prod DB after the build and then do a config export.

**Comment by mgamshad** [comment]
Regarding my note above:

> I am not sure if something is over writing the profile name during the sandbox site install, but if I change `core.extension.yml` to have the profile name from the error message then the error is cleared.

The above does seems to be coming from the profile definition in `blt.yml`.  In that file the install profile is listed as `[name]` and that's what the BLT `drush site-install` is using when installing the site during `blt setup`.  

Strange thing is why the production configuration is different than the BLT based local install?  We use `blt deploy` in a CI pipeline to push artifacts upstream.  Is BLT not using the profile named in `blt.yml`?

**Comment by mgamshad** [comment]
I did try loading the prod DB once `blt setup` was done and local sandbox was built.  Issuing `blt drupal:config:import` does succeed without any changes to the config files.  

However right after that, if I try from the top and re-run `blt setup`, even though the exported configurations now matches one-for-one (I did a complete folder to folder diff), `blt setup` fails during configuration import:

> The import failed due to the following reasons:                          
Entities exist of type **File**, **File type** **Image**. These entities need to be deleted before importing.  Cannot change the install profile from **`[name]`** to **lightning** once Drupal is installed. 

Unable to uninstall the **`[name]`** profile since it is the main install profile.

**Comment by mgamshad** [comment]
So I am getting my heard around what's going on, and I think at the end of the day the question on is on how to move an ACSF application from a Features based config management with BLT to one that uses core CMI and `config-split`.  And the main issue as I see it is with the BLT workflow.  

When `blt setup` is executed, the drupal site installation task will create a brand new site, and this site's UUID as well as all the other configuration object UUIDs will be different than the upstream remote ACSF site that is used from `blt.yml`.  This does not result in any config import issues when cm strategy is Features, since all of that UUID handling is taken care of by Features.  

In this same workflow, when we execute `blt sync:refresh`, we now get a set of configurations that match when compared upstream remote ACSF site.  The only slight differences is in some of the non-prod configurations that are specified in `blt.yml` and applied during this step.  However, again since Features is in control we have no config import issues.

I should also note that we have a handful of configuration files in the default sync folder (i.e. `[project-root]/config/default`) committed to code without any UUID assignments.  Given that the "Features" based BLT cm strategy is using the `--partial` flag on import, and combined with using Features, the import of these configurations does not cause any issues.

So now to move all of the above Featured based configuration to core CMI and `config-split` is a bit challenging within the context of the BLT workflow.  The ACSF application has a single code base for **all** of the sites that are part of an ACSF application.  So, if I am committing to code that platform level configuration for all the site, i.e. in `[project-root]/config/default`, those configurations can not have any UUIDs.  

Basically need to know from the BLT perspective what we need to do to move this project from Features to core CMI and config split.  (and really the main one is just core CMI, and not even worrying about any specific split issues).

Thanks.

**Comment by mikemadison13** [comment]
I've spent some time reviewing notes of when I did this with Features last time and I misspoke earlier about the UUID issue. I DO remember now having issues similar to this. BLT "works around" the UUID issue in general, but you may need to strip "specific" UUIDs from these failures out, and then re-export content.

Something about the features UUIDs is REALLY wonky. Once you get out of features, this shouldn't be a problem. I wouldn't nuke "all" the UUIDs but you should be able to run `drush cim` get a failure, nuke a uuid for that specific config item, and iterate until it goes cleanly (then export the config). From there, `blt setup` should be ok.

Sorry for the confusion! It's been a few years since I've done this.

**Comment by danepowell** [comment]
It seems like this discussion is wrapping up, is there some specific change we should make to the documentation to summarize the outcome, or to help other users with similar concerns?

**Comment by mgamshad** [comment]
I can come back and update when we have a way of handling the UUID issue.  I do think that the BLT documentation should include both this use case, i.e. moving off of Features to core CMI, and if nothing, to mention the UUID issue.  

The other area where we really need more documentation from BLT is on how to get the multi-site on ACSF use case implemented.  That is, if I have a set of sites on ACSF that I want to manage configuration for, and would like to create config-splits for, what setup steps are required, and what if any `blt deploy` command considerations are to be made?

**Comment by mgamshad** [comment]
So, I ignoring the split question here for multi-site for a moment, the issue here: [Allow site to be installed from existing config.](https://github.com/acquia/blt/issues/2565), is probably more related to what I am trying to do.  which is to take a BLT project that was configured to use `Features` to move to using core CMI for an existing set of sites on ACSF managed via BLT.

As for the UUIDs, I removed them for all except the `system.site.yml`, since I am going to use `config-split` to handle each site's UUID.  From the issue noted above, I noticed that I need to set `cm.core.install_from_config` to true in `blt.yml`.  Which sounds like it's what's needed to allow `blt setup` to install from an existing config.  I also had to remove the install hook from our custom Lightning-based profile. 

I also exported the configuration from a remote ACSF site that would be considered platform-level settings to my local  in the default sync directory.  BLT is picking up my flag and the default directory is detected.  At this point `blt setup` stops with following error:

>   The Drupal\taxonomy\Entity\Term class does not correspond to an entity type.

Trying to debug this now.  Any ideas?

Thanks

**Comment by mgamshad** [comment]
Reading this: [Support the "install site from existing config" feature of 8.6](https://github.com/acquia/blt/issues/3086) I wonder if my issue is that I should perhaps **not** be running `blt sync:refresh` if I am using `blt setup` and installing from an existing config?

**Comment by mgamshad** [comment]
On second thought `blt sync:refresh` should also be able to work as intended.  Right now I am able to get `blt setup` to complete, but with the message about the active and DB configuration mismatch.

**Comment by mgamshad** [comment]
With UUIDs removed, `blt setup` and `blt sync:refresh` do run, and config import is also successful, but BLT reports an error regarding the mismatch between what was imported and source configs.  But that seems to be BLT's second validation, which of course won't match as the UUIDs were stripped from the files.  I can confirm this by just running `drush ci` and it runs with no additional errors reported.

So right now if I run all the drush commands that BLT runs, I get a successful local built using BLT commands.  Any thoughts on this double-validation check that BLT runs in the context of having to mess with UUIDs?

Thanks

**Comment by mgamshad** [comment]
Ah, the last piece was this:
`allow-overrides: true`

Now I have both `blt setup` and `blt sync:refresh` working.  I do need to double check whether the few remaining UUIDs won't cause an issue, i.e. those that are not config object level IDs.

And finally I'd also like to sort out addressing the following warnings logged from `blt setup`:

>  [notice] Starting Drupal installation. This takes a while.
 [info] Undefined index: AH_SITE_GROUP AcsfSite.php:212
 [info] Undefined index: AH_SITE_ENVIRONMENT AcsfSite.php:213
 [warning] file_get_contents(/mnt/files/./nobackup/sf_shared_creds.ini): failed to open stream: No such file or directory AcsfSite.php:215

But apart from a round of testing, and then off to getting splits to work, we should be good.

**Comment by mgamshad** [comment]
Here is a recap to save time re-reading this:

Started with BLT project using Features and a partial set of configs in config/default folder.  The project code base is to support a multi-site platform on Acquia ACSF environment.

- Ran `blt setup` and `blt sync:refresh` to finish building the above project
- Actually synced the local to one of the ACSF remote sites in PROD
- Ran `druch cex -y` to export the default configuration
- Ran a script to strip the object level UUIDs and the `_core` key and its hash config value
- In the project's BLT configuration file (blt.yml) changed: `cm.stragety: config-split`, `cm.core.install_from_config: true`, and `cm.allow_overrides: true`
- Re-ran `blt setup` and `blt sync:refresh` to make sure I can rebuild sandbox

I'll update back here once I have the splits also tested and working.

**Comment by shamsher327** [comment]
@mgamshad which script do i need to run remove -Ran a script to strip the object level UUIDs and the _core key and its hash config value ?? Please guide

**Comment by deeplfo** [comment]
btw, if anyone interested, I used the following script to strip object-level UUIDs:

`
#!/bin/sh

# Stop if any command fails.
set -e

cd config/default; sed -i.bak '/^uuid: /d' ./*.yml; rm *.bak; cd - 
cd config/default; sed -i.bak '/default_config_hash: /d' ./*.yml; rm *.bak; cd - 
cd config/default; sed -i.bak '/^_core:/d' ./*.yml; rm *.bak; cd - 
`

**Comment by shamsher327** [comment]
it's works for me :) 

> btw, if anyone interested, I used the following script to strip object-level UUIDs:
> 
> `#!/bin/sh`
> `set -e`
> `cd config/default; sed -i.bak '/^uuid: /d' ./*.yml; rm *.bak; cd - `
> `cd config/default; sed -i.bak '/default_config_hash: /d' ./*.yml; rm *.bak; cd - `
> `cd config/default; sed -i.bak '/^_core:/d' ./*.yml; rm *.bak; cd - `

It's works for me
